File: backend/models/evidence.py
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import date
from enum import Enum
from typing import Optional


class SourceType(str, Enum):
    OFFICIAL = "OFFICIAL"                
    FEDERAL_REGULATOR = "FEDERAL_REGULATOR"  
    STATE_GOVERNMENT = "STATE_GOVERNMENT"    
    OPERATOR = "OPERATOR"               
    COMMUNITY = "COMMUNITY"              
    NEWS = "NEWS"                       
    RESEARCH = "RESEARCH"             
    SATELLITE_CONTEXT = "SATELLITE_CONTEXT"  


class ClaimField(str, Enum):
    OCCURRENCE = "OCCURRENCE"
    LOCATION = "LOCATION"
    DATE = "DATE"
    CAUSE = "CAUSE"
    VOLUME = "VOLUME"


class FieldStatus(str, Enum):
    CORROBORATED = "CORROBORATED"         
    SUPPORTED = "SUPPORTED"               
    CONTESTED = "CONTESTED"                
    CONFLICTING = "CONFLICTING"            
    UNRESOLVED = "UNRESOLVED"              
    NOT_ESTABLISHED = "NOT_ESTABLISHED"   
    NOT_ENOUGH_EVIDENCE = "NOT_ENOUGH_EVIDENCE"


@dataclass(frozen=True)
class Source:
    id: str
    name: str
    type: SourceType
    url: Optional[str] = None
    attributed_to: Optional[str] = None


@dataclass(frozen=True)
class Claim:
    id: str
    incident_id: str
    field: ClaimField
    raw_value: str
    source_id: str
    event_date: Optional[date] = None       
    publication_date: Optional[date] = None  
    excerpt: Optional[str] = None            
    derived_from_claim_id: Optional[str] = None


@dataclass
class ValueGroup:
    """One cluster of claims that agree with each other after normalization."""
    normalized_value: str
    display_value: str
    claim_ids: list[str] = field(default_factory=list)
    source_ids: set[str] = field(default_factory=set)  
    independent_source_ids: set[str] = field(default_factory=set)


@dataclass
class FieldAssessment:
    field: ClaimField
    status: FieldStatus
    groups: list[ValueGroup] = field(default_factory=list)
    contributing_claim_ids: list[str] = field(default_factory=list)
    note: str = ""
    reason: str = ""


@dataclass
class IncidentAssessment:
    incident_id: str
    fields: dict[ClaimField, FieldAssessment]
    summary: str

File: frontend/src/types/evidence.ts
export type SourceType =
  | "OFFICIAL"
  | "FEDERAL_REGULATOR"
  | "STATE_GOVERNMENT"
  | "OPERATOR"
  | "COMMUNITY"
  | "NEWS"
  | "RESEARCH"
  | "SATELLITE_CONTEXT";

export type ClaimField = "OCCURRENCE" | "LOCATION" | "DATE" | "CAUSE" | "VOLUME";

export type FieldStatus =
  | "CORROBORATED"
  | "SUPPORTED"
  | "CONTESTED"
  | "CONFLICTING"
  | "UNRESOLVED"
  | "NOT_ESTABLISHED"
  | "NOT_ENOUGH_EVIDENCE";

export interface Source {
  id: string;
  name: string;
  type: SourceType;
  url: string | null;
  attributed_to: string | null;
}

export interface Claim {
  id: string;
  field: ClaimField;
  raw_value: string;
  normalized_value: string | null;
  source_id: string;
  event_date: string | null;
  publication_date: string | null;
  excerpt: string | null;
  derived_from_claim_id: string | null;
  derived_from_label: string | null;
}

export interface ValueGroup {
  normalized_value: string;
  display_value: string;
  claim_ids: string[];
  source_ids: string[];
  independent_source_ids: string[];
}

export interface FieldAssessment {
  status: FieldStatus;
  note: string;
  narrative: string;
  reason: string;
  groups: ValueGroup[];
  contributing_claim_ids: string[];
}

export interface CasePayload {
  label: string;
  incident: {
    id: string;
    title: string;
    location_label: string;
    period_label: string;
  };
  sources: Source[];
  claims: Claim[];
  fields: Record<ClaimField, FieldAssessment>;
  summary: string;
}
